server/routerlicious/packages/services-core/src/lambdas.ts

```diff
@@ -165,6 +165,16 @@ export interface IPartitionLambdaConfig {
 	documentId: string;
 }
 
+/**
+ * Whether the boxcar message includes the optional Routing Key fields.
+ * @internal
+ */
+export function isCompleteBoxcarMessage(
+	boxcar: IBoxcarMessage,
+): boxcar is Required<IBoxcarMessage> {
```

Comment:
is this legal return type? maybe educate me how would this return type work?

Reply:
Yes, we use it in a lot of places in the codebase. It is super helpful for [Type Narrowing](https://www.typescriptlang.org/docs/handbook/2/narrowing.html) or [Type Guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards). It's similar to returning `boolean`, but it allows the TS compiler to know that the variable passed as `paramName` (`boxcar`, in this case) is actually of the indicated type. Basic example:

```ts
interface IInput {
  message: string;
}

function isValidInput(obj: unknown): obj is IInput {
  return typeof obj === "object" && obj !== null && message in obj && typeof obj.message === "string";
}

const userInput: unknown = { message: "hello" };
console.log(userInput.message); // Not OK
if (isValidInput(userInput)) {
  console.log(userInput.message); // OK
}
```